accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+# Unified Collections
+
+The Dart team is concurrently working on three proposals that affect collection
+literals:
+
+* [Set Literals][]
+* [Spread Collections][]
+* [Control Flow Collections][]
+
+[set literals]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/set-literals/feature-specification.md
+[spread collections]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/spread-collections/feature-specification.md
+[control flow collections]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/control-flow-collections/feature-specification.md
+
+These features interact in several ways, some of which weren't obvious at first.
+To make things easier on implementers and anyone else trying to understand the
+entire set of changes, this specification unifies and subsumes all three of
+those proposals.
+
+**This document is now the source of truth for these language changes.** The
+other three proposals are useful because they contain motivation and other
+context, but the precise syntax and semantics may be out of date in those docs.
+This is where you should be looking if you're an implementer.
+
+## Grammar
+
+Remove `setLiteral` and `mapLiteral` entirely. Then change the grammar to:
+
+```
+setOrMapLiteral   : 'const'? typeArguments? '{' elements? '}' ;
+listLiteral       : const? typeArguments? '[' elements? ']' ;
+
+elements          : element ( ',' element )* ','? ;
+
+element           : expressionElement
+                  | mapEntry
+                  | spreadElement
+                  | ifElement
+                  | forElement
+                  ;
+
+expressionElement : expression ;
+mapEntry          : expression ':' expression ;
+
+spreadElement     : ( '...' | '...?' ) expression ;
+ifElement         : 'if' '(' expression ')' element ( 'else' element )? ;
+forElement        : 'await'? 'for' '(' forLoopParts ')' element ;
+```
+
+Let *leaf elements* be all of the `expressionElement` and `mapEntry` elements in
+*e*, including elements of `ifElement` or `forElement` elements, transitively.
+
+It is a compile-time error if any *leaf elements* of a `listLiteral` are
+`mapEntry` elements. *(We could avoid this prose by duplicating the above rules
+for lists and removing `mapEntry`, but this is simpler.)*
+
+The existing rule that an expression statement cannot start with `{` now covers
+both map and set literals.
+
+## Static Semantics
+
+### Disambiguating sets and maps
+
+Because sets and maps both use curly braces as delimiters, you can create
+collection literals that are not obviously either a set or a map, like:
+
+```dart
+var a = {};
+var b = {...other};
+```
+
+When possible, we first use syntax and the surrounding context type to
+disambiguate between a map and set:
+
+Let *e* be a `setOrMapLiteral`.
+
+1.  If *e* has `typeArguments` then:
```

Comment:
This is unfortunate from a recovery point of view. It says (if I'm understanding it correctly) that the following code
```
var a = <int>{1 : 2, 2 : 4, 3 : 6, 4 : 8}
```
must be interpreted to be a set literal, which then causes the map entries to each be a compile time error. Tools will presumably produce four diagnostics of the form "you can't put a map entry in a set literal". But I think the more appropriate behavior would be to produce a single diagnostic of the form "map literals must have either zero or two type arguments".

Any chance we can rework this algorithm to recover better in the face of mismatches between the number of type arguments and the kind of elements in the collection?

Reply:
As I understand it, the spec and the algorithm doesn't force you to produce any particular set of errors. All it says is that it "is a compile-time error" for the literal to have certain properties. I don't think we take the "a" part of that literally. We could probably just as well word it "it is erroneous".

As long as at least one error is reported, somewhere, I think the proposal and spec are happy. @lrhn @leafpetersen, do you agree?

For example:

```dart
<int>["not", "an", "int", "these", "are", "strings"];
```

Do you think the analyzer should/must report six type errors there instead of suggesting they fix the type argument?

Reply:
Absolutely agree. The program has a compile-time error. That just means that the program has no run-time semantics. The moment a compiler detects that anything is a compile-time error, it can stop right there and say "Program has a compile-time error", and the specification is satisfied. 
Everything on top of that is usability improvements, and you are free to do whatever is best for the use (short of accepting invalid programs or rejecting valid programs, at least without explicit user opt-in like the "no-implicit-downcast" flag).
Error recovery is a heuristic approach to make some sense of a program with a compile-time error, in order to be able to report more compile-time errors in the same run (and with cascade errors as the failure mode of that heuristic).

Most of the specification given here is of the form "if t1 then if t2. then success otherwise compile-time error", which you can read as the constraint: "t1 => t2". If that fails to be satisfied, it's not clear whether it's because of t1 or because of not t2. You can report either, or both, in order to tell the user what went wrong.
The example given here is one of having the failing implication (`<int>` prefix => must be set => must not contains map entry like `1: 2`).


There is some sequencing issues in the specification, where we speak about things in multiple phases (any text that talks about *types* assumes that a the class declarations define a valid type hierarchy, it makes no sense to talk about types and subtyping before ensuring that there are no cycles in the class declarations), but in general, we try to avoid saying that one static check happens before another.

Reply:
Excellent! Thanks for the clarification.